### PR TITLE
[FLINK-27146] [Filesystem] Migrate to Junit5

### PR DESCRIPTION
**Pull request by kottmann**
## What is the purpose of the change

This migrates the Filesystem test code to use Junit5 and Assertj. 

## Brief change log

- Migrated the test code to use Junit5 according to the migration documentation
- Also migrated to Assertj (this is in a separate commit and could be dropped or resubmitted in a separate PR)

## Verifying this change

This change will not have any impact on non-test code since only test code was changed. The  change can potentially break tests, but the majority of changes are simple and don't touch the logic.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopFsFactoryTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopFsFactoryTest.java:55]: Use AssertJ's `assertThatThrownBy` instead of try/catch.

File: flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopFreeTests.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopFreeTests.java:1]: Use AssertJ's `assertThatThrownBy` instead of try/catch and `fail()`.

File: flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemITCase.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemITCase.java:36]: Use AssertJ's `assertThatThrownBy` instead of try/catch `fail()`.

**Review by ferenc-csaky — CHANGES_REQUESTED** [review]
Thank you for your contribution! If you have some time, pls. review my comments.

In general, pls. use AssertJ for any kind of asserts. The community agreed on this and it is part of the [JUnit5 migration plan](https://docs.google.com/document/d/1514Wa_aNB9bJUen4xm5uiuXOooOJTtXqS_Jqk9KJitU/edit#heading=h.n2pt1kpwyals). Currently, there are JUnit5 asserts, some `junit.framework.TestCase.assert*`-like imports (this style is a JUnit3 `TestCase` extension residue), and `org.hamcrest.MatcherAssert.assert*`-like imports. All of those should be replaced by AssertJ assertions.

Some test classes and methods were changed to package-private, some are not. Pls. make those changes on all test classes.

**Comment by kottmann** [comment]
@ferenc-csaky  Thank you for your insightful feedback and for directing me towards the assertj decision. I updated my branch by rebasing it onto master and incorporating the changes related to assertj. Additionally, I will make another commit to address the remaining feedback that isn't connected to assertj.

File: flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/utils/ConfigUtilsHadoopTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/utils/ConfigUtilsHadoopTest.java:285]: Use `.isEmpty()` instead of equal check.

File: flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableFsDataOutputStreamTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableFsDataOutputStreamTest.java:138]: nit: Use `.isZero()`. Applies to every other zero-check in the file.

File: flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableFsDataOutputStreamTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableFsDataOutputStreamTest.java:169]: nit: Use `.isOne()`.

File: flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableWriterCommitterTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableWriterCommitterTest.java:140]: nit: Use `.isOne()`.

File: flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/AbstractHadoopFileSystemITTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/AbstractHadoopFileSystemITTest.java:106]: nit: Use `.isZero()`.

File: flink-filesystems/flink-oss-fs-hadoop/src/test/java/org/apache/flink/fs/osshadoop/writer/OSSRecoverableSerializerTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-oss-fs-hadoop/src/test/java/org/apache/flink/fs/osshadoop/writer/OSSRecoverableSerializerTest.java:112]: nit: static import for `assertThat`.

File: flink-filesystems/flink-s3-fs-base/src/test/java/org/apache/flink/fs/s3/common/writer/RecoverableMultiPartUploadImplTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-s3-fs-base/src/test/java/org/apache/flink/fs/s3/common/writer/RecoverableMultiPartUploadImplTest.java:145]: Both custom assert method with a for-loop can be simplified to something like this:
```java
private static void assertThatHasUploadedObject(StubMultiPartUploader actual, byte[] content) {
    TestPutObjectResult expectedIncompletePart =
            createPutObjectResult(TEST_OBJECT_NAME, content);

    assertThat(actual.getIncompletePartsUploaded()).contains(expectedIncompletePart);
}
```

File: flink-filesystems/flink-s3-fs-base/src/test/java/org/apache/flink/fs/s3/common/writer/S3RecoverableFsDataOutputStreamTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-s3-fs-base/src/test/java/org/apache/flink/fs/s3/common/writer/S3RecoverableFsDataOutputStreamTest.java:270]: nit: static import for `assertThatThrownBy`.

**Review by ferenc-csaky — CHANGES_REQUESTED** [review]
Thanks for going through the files again! Added some minor improvement suggestions, mostly nits about common practice in the repo.

**Comment by kottmann** [comment]
Sorry for the delay on my end, I added a new commit to address all your comments above.

**Review by ferenc-csaky — APPROVED** [review]
LGTM! @Jiabao-Sun if you have some time, can you check this PR?

File: flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableFsDataOutputStreamTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableFsDataOutputStreamTest.java:138]: This should be `.isZero()`, and causes the Azure fail. Missed that in my prev comment. Pls. fix this, if tests are green, then LGTM :)

**Review by ferenc-csaky — CHANGES_REQUESTED** [review]

**Review by ferenc-csaky — APPROVED** [review]

**Comment by kottmann** [comment]
@ferenc-csaky Thank you for reviewing my PR and for your helpful feedback! I really appreciate your support in improving it.

**Comment by ferenc-csaky** [comment]
@kottmann I was not a committer back then, but if you will have some time and desire to resolve the current conflicts, I still believe this would be a nice improvement and will add it afterwards.

**Comment by Jiabao-Sun** [comment]
Thanks @kottmann for this contribution and @ferenc-csaky for the review, 
Sorry for missing the message earlier.
Could you rebase master and resolve the conflicts?

**Comment by kottmann** [comment]
Thanks, I rebased it onto master and it should be possible to merge now. Some classes received a bit of JUnit 5 migration from another commit before, and I resolved those conflicts or removed my changes.

**Review by ferenc-csaky — APPROVED** [review]
Thanks for updating, LGTM, will merge in 24h if no objections until then.

File: flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopRecoverableWriterTest.java
Comment (ferenc-csaky) [reply flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopRecoverableWriterTest.java:51]: The actual min Hadoop version is 2.10 for a while: https://github.com/apache/flink/blob/adb8368985ad20e1872a50939c43488516d18b55/pom.xml#L113

So effectively both will be true, it won't matter.

File: flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/GSFileSystemFactoryTest.java
Comment (Jiabao-Sun) [inline flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/GSFileSystemFactoryTest.java:31]:
```suggestion
   void testOverrideStorageRootUrl() {
```

File: flink-filesystems/flink-oss-fs-hadoop/src/test/java/org/apache/flink/fs/osshadoop/HadoopOSSRecoverableWriterExceptionITCase.java
Comment (Jiabao-Sun) [inline flink-filesystems/flink-oss-fs-hadoop/src/test/java/org/apache/flink/fs/osshadoop/HadoopOSSRecoverableWriterExceptionITCase.java:47]:
```suggestion
   static void checkCredentialsAndSetup() throws IOException {
```

Comment (Jiabao-Sun) [inline flink-filesystems/flink-oss-fs-hadoop/src/test/java/org/apache/flink/fs/osshadoop/HadoopOSSRecoverableWriterExceptionITCase.java:47]:
```suggestion
   static void checkCredentialsAndSetup() throws IOException {
```

File: flink-filesystems/flink-oss-fs-hadoop/src/test/java/org/apache/flink/fs/osshadoop/HadoopOSSRecoverableWriterITCase.java
Comment (Jiabao-Sun) [inline flink-filesystems/flink-oss-fs-hadoop/src/test/java/org/apache/flink/fs/osshadoop/HadoopOSSRecoverableWriterITCase.java:49]:
```suggestion
   static void checkCredentialsAndSetup() throws IOException {
```

File: flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemITCase.java
Comment (Jiabao-Sun) [inline flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemITCase.java:47]: Will we lose parameter tests? Additionally, can we check if the number of tests is consistent with the master branch?

**Review by Jiabao-Sun — COMMENTED** [review]
Thanks @kottmann for this contribution and @ferenc-csaky for the review.
I left some comments.

File: flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemITCase.java
Comment (kottmann) [reply flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemITCase.java:47]: The parameter "schema" wasn't used in the test before, therefore nothing should change by removing it. 

Let me compare the number of tests and then comment here.

**Comment by kottmann** [comment]
@Jiabao-Sun Thanks for your review, I addressed your comments in the last commit, for the more general things you mentioned I searched in flink-filesystem for other occurrence of the same pattern and changed it too.

I am not sure about the number of tests running, it appears that the number of tests in CI for my branch is lower than for master, and also the test duration is shorter, so it would be good to understand whats going on before merging this.

**Comment by kottmann** [comment]
I tracked a missing test case down to HadoopS3FileSystemBehaviorITCase this test doesn't run on my branch, but does run on the master branch. But this wasn't modified in my PR, do you know why this doesn't run? Are they disabled by default in CI for PRs?

**Review by Jiabao-Sun — COMMENTED** [review]
Thanks @kottmann for the quick update.
Overall, it was good for me. I just left a few comments. 
Also, did the CI fail this time because of the changes made?

File: flink-filesystems/flink-azure-fs-hadoop/src/test/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriterTest.java
Comment (ferenc-csaky) [inline flink-filesystems/flink-azure-fs-hadoop/src/test/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriterTest.java:53]: I think we should revert this part to AssertJ's `aasumeThat`.

File: flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopRecoverableWriterTest.java
Comment (ferenc-csaky) [reply flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopRecoverableWriterTest.java:51]: Although, I think we should keep using AssertJ's `assumeThat`.

**Review by ferenc-csaky — CHANGES_REQUESTED** [review]
Thank you @Jiabao-Sun for your thorough review! Added some minor comments myself. 

The force-push compare got me tricked, cause it showed only a couple line changes and I thought my last review is still intact. :)

**Review by ferenc-csaky — APPROVED** [review]
I'm pretty sure that the last failure cause by that timeout in the "connect" test module is completely independent from your changes. Either a flaky test or some other merged commit to master.

Thank you for your responsiveness and quick fixes! Checked through all the changed files, and in overall the current state LGTM. I will merge it tomorrow if no objections until then. cc @Jiabao-Sun

**Comment by kottmann** [comment]
@ferenc-csaky Shall I squash the commits, or will you do that when it gets merged?

**Comment by ferenc-csaky** [comment]
> @ferenc-csaky Shall I squash the commits, or will you do that when it gets merged?

Squash and merge takes care of that, so it is not necessary I would say.

**Review by Jiabao-Sun — APPROVED** [review]
Thanks @kottmann for the great work and @ferenc-csaky for the review.
LGTM now.

**Comment by kottmann** [comment]
@ferenc-csaky  @Jiabao-Sun  Thank you two for your helpful review. If there are any more tickets related to the JUnit 5 migration, I can help with I am happy to take one over.

**Pull request merged**
